mod/metric_answer_lookup/lib/answer/cardless_answers.rb

```diff
@@ -0,0 +1,66 @@
+class Answer
+  # MEthods to handle answers that exist only in the the answer table
+  # and don't have a card. Used for calculated answers.
+  module CardlessAnswers
+    def self.included host_class
+      host_class.extend ClassMethods
+    end
+
+    def virtual_answer_card name=nil, val=nil
+      name ||= [record_name, year.to_s]
+      val ||= value
+      Card.new(name: name, type_id: Card::MetricValueID).tap do |card|
+        card.define_singleton_method(:value) { val }
+        # card.define_singleton_method(:updated_at) { updated_at }
+        card.define_singleton_method(:value_card) do
+          Card.new name: [name, :value], content: val
+        end
```

Comment:
dunno if you saw my outdated comment on this:
> we might consider adding all these virtual card methods to a module and having the singleton_class include it. My hunch is that it would be faster (performance-wise)

I don't want to push for that, but I am curious to hear your rationale for this approach.

Reply:
the rationale is less lines of code and faster:
```
class A; end

def with_module val
  mod = Module.new do
    define_method :x do
      puts val
    end
    define_method :y do
      puts val
    end
  end
  a = A.new
  a.singleton_class.include mod
  a.x
end

def with_method val
  a = A.new
  a.define_singleton_method :x do
    puts val
  end
  a.define_singleton_method :y do
    puts val
  end
  a.x
end
```
Comparison:
         with method:    51961.0 i/s
         with module:    37984.6 i/s - 1.37x  slower

Reply:
But in the test you're defining the module over and over again; that's definitely going to be slower.  I was suggesting that you define the module once and then include it over and over again.  I don't _really_ know, but I'd predict that would be an order of magnitude faster.

Reply:
I don't see how I can define module only once. The definition of the module depends on the argument of the method that includes the module.

Reply:
I could do it like this
```
class A; end

module M
  attr_accessor :val1, :val2
  def x; puts val1 end
  def y; puts val2 end
end

def with_single_module val1, val2
  a = A.new
  a.singleton_class.include M
  a.val1 = val1
  a.val2 = val2
  a.x
  a.y
end
```
but it makes no difference:

Comparison:
with dymanical;y defined methods:    60403.9 i/s
with hard-coded module:    59385.8 i/s - same-ish: difference falls within error
with dynamically defined module:    49653.3 i/s - 1.22x  slower

Reply:
the only way I found so far to get it faster is if the module is initially prepended: 
```ruby
class A
  def x
    puts "original method"
  end
end

module BM
  attr_accessor :val1, :val2
  def x
    return super unless val1
    puts val1
  end
end


class B < A
  prepend BM
end
```
Including the module at runtime is 1.25x slower than using `B` like this
```ruby
a = B.new
a.val1 = val1
a.x
```

Reply:
if you want to try yourself: https://gist.github.com/xithan/dd6b77c8ca88b99e9f27da21cd37b1e2

Reply:
OK, I get it now and see why you've done it this way.  Sorry, should have studied a bit closer.  Thank you!